qiskit/algorithms/minimum_eigensolvers/vqe.py

```diff
@@ -0,0 +1,333 @@
+
+        Args:
+            ansatz: The ansatz preparing the quantum state.
+            operator: The operator whose energy to evaluate.
+
+        Returns:
+            A function handle to evaluate the gradient at given parameters for the ansatz.
+
+        Raises:
+            AlgorithmError: If the primitive job to evaluate the gradient fails.
+        """
+
+        def evaluate_gradient(parameters: np.ndarray) -> np.ndarray:
+            # broadcasting not required for the estimator gradients
+            try:
+                job = self.gradient.run([ansatz], [operator], [parameters])
+                gradients = job.result().gradients
+            except Exception as exc:
+                raise AlgorithmError("The primitive job to evaluate the gradient failed!") from exc
+
+            return gradients[0]
+
+        return evaluate_gradient
+
+    def _check_operator_ansatz(self, operator: BaseOperator | PauliSumOp):
+        """Check that the number of qubits of operator and ansatz match and that the ansatz is
+        parameterized.
+        """
+        if operator.num_qubits != self.ansatz.num_qubits:
+            try:
+                logger.info(
+                    "Trying to resize ansatz to match operator on %s qubits.", operator.num_qubits
+                )
+                self.ansatz.num_qubits = operator.num_qubits
+            except AttributeError as error:
+                raise AlgorithmError(
+                    "The number of qubits of the ansatz does not match the "
+                    "operator, and the ansatz does not allow setting the "
+                    "number of qubits using `num_qubits`."
+                ) from error
+
+        if self.ansatz.num_parameters == 0:
+            raise AlgorithmError("The ansatz must be parameterized, but has no free parameters.")
+
+    def _build_vqe_result(
+        self,
+        optimizer_result: OptimizerResult,
+        aux_values: ListOrDict[tuple[complex, tuple[complex, int]]],
+        eval_time: float,
+    ) -> VQEResult:
+        result = VQEResult()
+        result.eigenvalue = optimizer_result.fun
+        result.cost_function_evals = optimizer_result.nfev
+        result.optimal_point = optimizer_result.x
+        result.optimal_parameters = dict(zip(self.ansatz.parameters, optimizer_result.x))
+        result.optimal_value = optimizer_result.fun
+        result.optimizer_time = eval_time
+        result.aux_operator_eigenvalues = aux_values
+        result.optimizer_result = optimizer_result
+        return result
+
+
+class VQEResult(VariationalResult, MinimumEigensolverResult):
+    """Variational quantum eigensolver result."""
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._cost_function_evals = None
+
+    @property
+    def cost_function_evals(self) -> int | None:
+        """The number of cost optimizer evaluations."""
+        return self._cost_function_evals
```

Comment:
I guess this is fine so treat this comment more as just an observation. This used to be an eval count that VQE made - but over time it seems to have ended up with what the optimizer said it did - ie nfev. Pretty much all the optimizer related result values are in VariationalResult and as such nowadays, with this being from the optimizer, it could be pretty much in that one place - other algos like VQD then would not need to define it in their result. At this stage its fine - we could move it up one level in the future without breaking things.